guides/20240820_guide_devcontainer_feature.md

```diff
@@ -0,0 +1,272 @@
+    },
+    "tokenizers_version": {
+      "type": "string",
+      "proposals": ["latest", "0.11.6", "0.11.5"],
+      "default": "latest",
+      "description": "Select the version of Tokenizers to install."
+    }
+  },
+  "installsAfter": ["ghcr.io/devcontainers/features/common-utils"]
+}
+```
+
+**Note:** _[Ensure that descriptions are clear and informative, making it easy for users to understand what each option does. This step is crucial for providing a seamless user experience]_
+
+### Implement `install.sh`
+
+The `install.sh` script is the core of the feature’s installation process. Below is an optimized version of the script that incorporates modular functions and error handling.
+
+```bash
+#!/bin/bash
+set -e
+
+echo "Activating feature 'huggingface'"
+
+# Function to install Python packages
+install_python_package() {
+    local package_name=$1
+    local package_version=$2
+    if [ "$package_version" = "latest" ]; then
+        pip install $package_name
+    else
+        pip install "$package_name==$package_version"
+    fi
+}
+
+# Ensure Python and pip are available
+if ! command -v python3 &> /dev/null; then
+    apt-get update && apt-get install -y python3 python3-pip python3-venv
+else
+    echo "Python3 is already installed."
+fi
+
+# Set up a virtual environment if not present
+VENV_PATH="/opt/huggingface-venv"
+if [ ! -d "$VENV_PATH" ]; then
+    python3 -m venv $VENV_PATH
+    echo "Virtual environment created at $VENV_PATH."
+else
+    echo "Using existing virtual environment at $VENV_PATH."
+fi
+
+# Activate the virtual environment
+source $VENV_PATH/bin/activate
+
+# Upgrade pip to the latest version
+pip install --upgrade pip
+
+# Install required Python packages
+install_python_package "transformers" "$VERSION"
+
+# Install PyTorch with or without CUDA support
+if [ "$CUDA" = "true" ]; then
+    pip install torch torchvision torchaudio
+else
+    pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu
+fi
+
+install_python_package "datasets" "$DATASETS_VERSION"
+install_python_package "tokenizers" "$TOKENIZERS_VERSION"
+pip install sentencepiece huggingface_hub
+
+# Create a script to activate the virtual environment
+cat <<EOL > /usr/local/bin/activate-huggingface
+#!/bin/bash
+source $VENV_PATH/bin/activate
+EOL
+chmod +x /usr/local/bin/activate-huggingface
+
+echo "Hugging Face setup completed successfully."
+```
+
+**Note:** _[The script allows users to customize their installation through the provided options, making the feature flexible for various use cases]_
+
+## Step 3: Confirmation
+
+After implementing the feature, confirm that everything works as expected by testing the feature locally. For testing use GitHub Actions to automatically test the feature. You can also test locally using the Dev Container CLI:
+
+```bash
+devcontainer features test -f huggingface -i mcr.microsoft.com/devcontainers/base:ubuntu .
```

Comment:
@nkkko Can I add a basic test to check their versions? Because I have seen this cd expects tests dir ?

Reply:
Added